### PR TITLE
qemu_guest_agent: Add case of umount fs while fsfrozen

### PR DESCRIPTION
**Pull request by xiagao**
New case of umount fs when fs is frozen
ID: 1726941
Signed-off-by: Xiaoling Gao <xiagao@redhat.com>

### CODE REVIEW

**Comment by vivianQizhu** [comment]
@xiangchunfu Request for your review, thanks.

File: qemu/tests/qemu_guest_agent.py [hunk 1]
```diff
@@ -1828,6 +1830,97 @@ def gagent_check_nonexistent_cmd(self, test, params, env):
...
+        try:
+            self.gagent.fsthaw()
+        except guest_agent.VAgentCmdError as detail:
+            if not re.search("fsfreeze is limited up to 10 seconds", str(detail)):
+                test.error("guest-fsfreeze-thaw cmd failed with: ('%s')"
+                           % str(detail))
+        error_context.context("Mount fs or online disk in guest.", logging.info)
+        if params.get("os_type") == "linux":
+            if not utils_disk.mount(partition, mnt_point[0], session=session):
+                test.fail("Can't mount fs with fsthaw.")
```
Comment (xiangchunfu) [inline qemu/tests/qemu_guest_agent.py]: @xiagao For different guest. It's behavior is different.  e.g  RHEL6 it should fail. RHEL7 and RHEL8 it should pass.

File: qemu/tests/qemu_guest_agent.py [hunk 2]
```diff
@@ -1849,6 +1853,124 @@ def gagent_check_with_migrate(self, test, params, env):
...
+
+        error_context.context("Umount fs or offline disk in guest after"
+                              " fsfreeze.", logging.info)
+        if params.get("os_type") == "linux":
+            # for rhel6 guest, umount will fail as timeout.
+            try:
+                umount_cmd = "umount %s" % mnt_point[0]
+                session.cmd_status(umount_cmd)
+            except ShellTimeoutError:
+                if is_rhel6 == "yes":
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py]: if `os_variants == 'rhel6'`

File: qemu/tests/qemu_guest_agent.py [hunk 3]
```diff
@@ -1849,6 +1853,124 @@ def gagent_check_with_migrate(self, test, params, env):
...
+                    logging.info("Umount cmd timeout for rhel6 guest as"
+                                 " expected.")
+                else:
+                    test.fail("Umount cmd timeout for rhel7+ guest.")
+            else:
+                test.fail("Umount fs failed after fsfreeze.")
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py]: Umount fs `should` failed after fsfreeze?

File: qemu/tests/qemu_guest_agent.py [hunk 4]
```diff
@@ -1849,6 +1853,124 @@ def gagent_check_with_migrate(self, test, params, env):
...
+            cmd_header = "echo list disk > " + disk
+            cmd_header += " && echo select disk %s >> " + disk
+            cmd_footer = " echo exit >> " + disk
+            cmd_footer += " && diskpart /s " + disk
+            cmd_footer += " && del /f " + disk
+            cmd_list = []
+            for i in cmd.split(";"):
+                i += " >> " + disk
+                cmd_list.append(i)
+            cmd = " && ".join(cmd_list)
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py]: `cmd = cmd.replace(';', '>> %s &&' disk) + '>> %s' % disk`? While actually I see no where has `;` for `cmd`, so directly `cmd += " >> %s" % disk` is enough?

File: qemu/tests/qemu_guest_agent.py [hunk 5]
```diff
@@ -1849,6 +1853,124 @@ def gagent_check_with_migrate(self, test, params, env):
...
+        is_rhel6 = params.get("is_rhel6")
+
+        error_context.context("Format the new data disk and mount it.",
+                              logging.info)
+        if params.get("os_type") == "linux":
+            disk_data = list(utils_disk.get_linux_disks(session).keys())
+            mnt_point = utils_disk.configure_empty_disk(
+                session, disk_data[0], image_size_stg0, "linux",
+                labeltype="msdos")
+            partition = "/dev/%s" % mnt_point[0].split("/")[-1]
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py]: I don't understand this usage, is `mnt_point[0]` like `/dev/sda` or directly `sda` something? Which case you need `split("/")[-1]` and add `/dev` before it?

File: qemu/tests/qemu_guest_agent.py [hunk 6]
```diff
@@ -1849,6 +1853,124 @@ def gagent_check_with_migrate(self, test, params, env):
...
+                120)
+            if disk_index:
+                logging.info("Clear readonly for disk and online it in"
+                             " windows guest.")
+                if not utils_disk.update_windows_disk_attributes(session,
+                                                                 disk_index):
+                    test.error("Failed to update windows disk attributes.")
+                mnt_point = utils_disk.configure_empty_disk(
+                    session, disk_index[0], image_size_stg0, "windows",
+                    labeltype="msdos")
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py]: Do nothing when no `disk_index`?

File: qemu/tests/qemu_guest_agent.py (hunk 3 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: For rhel6 guest,it will fail.

File: qemu/tests/qemu_guest_agent.py (hunk 4 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: You are right, modified

File: qemu/tests/qemu_guest_agent.py (hunk 5 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: Modified.

File: qemu/tests/qemu_guest_agent.py (hunk 6 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: add else operation.

File: qemu/tests/qemu_guest_agent.py [hunk 7]
```diff
@@ -1829,25 +1831,125 @@ def gagent_check_nonexistent_cmd(self, test, params, env):
...
+            cmd_footer += " && diskpart /s " + disk
+            cmd_footer += " && del /f " + disk
+            cmd += " >> " + disk
+            return " && ".join([cmd_header, cmd, cmd_footer])
+
+        session = self._get_session(params, self.vm)
+        image_size_stg0 = params["image_size_stg0"]
+
+        if params['os_variant'] == 'rhel8':
+            session.cmd("setenforce 0")
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py]: You might need update this later when we have newest release. I will prefer something like:
```
setenforce = True
Windows, RHEL.6, RHEL.7:
    setenforce = False
```
Or if it does not affect old release test, just send the command for all linux guest.

Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: I delete this part,as there is a product bug of rhel8, these code are just used to debug.
Thanks.

**Comment by xiagao** [comment]
depends on:
https://github.com/avocado-framework/avocado-vt/pull/2040/

**Comment by xiagao** [comment]
@xiangchunfu  I add rhel6 guest support,Could you have a check?

**Comment by xiagao** [comment]
(1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.6.10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_umount_frozen: PASS (141.35 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_umount_frozen: PASS (414.04 s)

**Comment by xiangchunfu** [comment]
Acked by: Xiangchun Fu xfu@redhat.com

**Comment by xiagao** [comment]
@vivianQizhu Call for your review..

File: qemu/tests/qemu_guest_agent.py [hunk 8]
```diff
@@ -1829,25 +1832,128 @@ def gagent_check_nonexistent_cmd(self, test, params, env):
             test.fail("Should return error info.")
 
     @error_context.context_aware
-    def gagent_check_with_migrate(self, test, params, env):
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py]: @xiagao Why delete this function?

Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: @vivianQizhu  I'm supposing it's branch conflict. And now it's back.

**Comment by vivianQizhu** [comment]
Ack

**Comment by vivianQizhu** [comment]
@xiagao Latest test result please, thanks.

**Comment by xiagao** [comment]
(1/3) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.6.10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_umount_frozen: PASS (130.75 s)
 (2/3) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_umount_frozen: PASS (38.70 s)
 (3/3) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_umount_frozen: PASS (124.81 s)

**Comment by vivianQizhu** [comment]
@xiagao Why remove the change for s/params.get()/params[]/g? And what's else updated after you post the test result?

**Comment by xiagao** [comment]
> @xiagao Why remove the change for s/params.get()/params[]/g? And what's else updated after you post the test result?

As check_snapshot test case will fail if keep params[], as there is no defination in it's config.
Will paste the result later, it's running now.

**Comment by xiagao** [comment]
(1/3) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.6.10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_umount_frozen: PASS (137.47 s)
 (2/3) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_umount_frozen: PASS (398.98 s)
 (3/3) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_umount_frozen: PASS (417.72 s)

 (1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_umount_frozen: PASS (51.97 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_umount_frozen: PASS (115.94 s)

**Pull request merged**